### PR TITLE
[AppSec] Automated user ID collection modes 

### PR DESCRIPTION
**Pull request by Anilm3**
### Description

This PR introduces the new user ID collection modes (`ident`, `anon`, `disabled`) which replace the old ones (`safe`, `extended`, `disabled`). Some interesting facts:
- `ident` mode is now the default one, meaning the user IDs are always collected.
- Metadata is no longer included on `ident` or `anon` mode.
- All relevant headers are now collected on `ident` or `anon` mode.
- Some of the functionality has been moved to `user_tracking.*`, in preparation for further work on authenticated user tracking.

The system tests will fail due to the change on login events, however they have been updated here: https://github.com/DataDog/system-tests/compare/anilm3/php-user-events-v2?expand=1

The remote configuration changes have not been included in this PR given the ongoing work being done on the helper.

Related Jira: [APPSEC-53749]

### Reviewer checklist
- [ ] Test coverage seems ok.
- [ ] Appropriate labels assigned.

[APPSEC-53749]: https://datadoghq.atlassian.net/browse/APPSEC-53749?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by estringana — APPROVED** [review]

File: appsec/src/extension/user_tracking.c [hunk 1]
```diff
@@ -45,6 +60,30 @@ static PHP_FUNCTION(set_user_wrapper)
...
+        LSTRARG("identification"), 1 /* persistent */);
+    _user_mode_anon_zstr =
+        zend_string_init_interned(LSTRARG("anonymization"), 1 /* persistent */);
+    _user_mode_disabled_zstr =
+        zend_string_init_interned(LSTRARG("disabled"), 1 /* persistent */);
+    _sha256_algo_zstr =
+        zend_string_init_interned(LSTRARG("sha256"), 1 /* persistent */);
+
+#if PHP_VERSION_ID < 80000
+    void *handle = dlopen(NULL, RTLD_NOW | RTLD_GLOBAL);
```
Comment (cataphract) [inline appsec/src/extension/user_tracking.c:73]: skip `dlopen` and use `dlsym` with `RTLD_DEFAULT`

File: appsec/src/extension/user_tracking.c [hunk 2]
```diff
@@ -115,3 +154,104 @@ void dd_find_and_apply_verdict_for_user(zend_string *nonnull user_id)
...
 }
+
+bool dd_parse_user_collection_mode(
+    zai_str value, zval *nonnull decoded_value, bool persistent)
+{
+    if (!value.len) {
+        return false;
+    }
+
+    size_t len = strlen((const char *)value.ptr);
```
Comment (cataphract) [inline appsec/src/extension/user_tracking.c:165]: isn't this what `value.len` is?

File: appsec/src/extension/user_tracking.c [hunk 3]
```diff
@@ -115,3 +154,104 @@ void dd_find_and_apply_verdict_for_user(zend_string *nonnull user_id)
...
+    if (dd_string_equals_lc(value.ptr, len, ZEND_STRL("ident")) ||
+        dd_string_equals_lc(value.ptr, len, ZEND_STRL("identification")) ||
+        dd_string_equals_lc(value.ptr, len, ZEND_STRL("extended"))) {
+        _user_mode = user_mode_ident;
+    } else if (dd_string_equals_lc(value.ptr, len, ZEND_STRL("anon")) ||
+               dd_string_equals_lc(
+                   value.ptr, len, ZEND_STRL("anonymization")) ||
+               dd_string_equals_lc(value.ptr, len, ZEND_STRL("safe"))) {
+        _user_mode = user_mode_anon;
+    } else { // If the value is disabled or an unknown value, we disable user ID
```
Comment (cataphract) [inline appsec/src/extension/user_tracking.c:175]: maybe should warn for an unknown value

File: appsec/src/extension/user_tracking.c [hunk 4]
```diff
@@ -115,3 +154,104 @@ void dd_find_and_apply_verdict_for_user(zend_string *nonnull user_id)
...
+    ops->hash_init(context);
+#else
+    ops->hash_init(context, NULL);
+#endif
+
+    ops->hash_update(
+        context, (unsigned char *)ZSTR_VAL(user_id), ZSTR_LEN(user_id));
+
+    digest = zend_string_alloc(ops->digest_size, 0);
+    ops->hash_final((unsigned char *)ZSTR_VAL(digest), context);
```
Comment (cataphract) [inline appsec/src/extension/user_tracking.c:222]: more of a comment on poor spec drafting, but this is very bad anonymization. You can just look up a rainbow table and figure out a significant portion of the usernames

File: appsec/tests/extension/headers_collection_10.phpt [hunk 5]
```diff
@@ -80,27 +80,57 @@ var_dump($headers);
 $helper->finished_with_commands();
 ?>
 --EXPECTF--
-array(11) {
-  ["http.request.headers.accept"]=>
-  string(3) "*/*"
+array(25) {
   ["http.request.headers.akamai-user-risk"]=>
   string(13) "akamaiuserisk"
+  ["http.request.headers.x-forwarded"]=>
```
Comment (cataphract) [inline appsec/tests/extension/headers_collection_10.phpt:86]: this should fail, as there's a ksort

**Review by cataphract — APPROVED** [review]

**Pull request merged**